crypto/merklearray/partial.go

```diff
@@ -0,0 +1,107 @@
+package merklearray
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/algorand/go-algorand/crypto"
+)
+
+// siblings represents the siblings needed to compute the root hash
+// given a set of leaf nodes.  This data structure can operate in two
+// modes: either build up the set of sibling hints, if tree is not nil,
+// or use the set of sibling hints, if tree is nil.
+type siblings struct {
+	tree  *Tree
+	hints []crypto.Digest
+}
+
+// get returns the sibling from tree level l (0 being the leaves)
+// position i.
+func (s *siblings) get(l uint64, i uint64) (res crypto.Digest, err error) {
+	if s.tree == nil {
+		if len(s.hints) > 0 {
+			res = s.hints[0]
+			s.hints = s.hints[1:]
+			return
+		}
+
+		err = fmt.Errorf("no more sibling hints")
+		return
+	}
+
+	if l >= uint64(len(s.tree.levels)) {
+		err = fmt.Errorf("level %d beyond tree height %d", l, len(s.tree.levels))
+		return
+	}
+
+	if i < uint64(len(s.tree.levels[l])) {
+		res = s.tree.levels[l][i]
+	}
+
+	s.hints = append(s.hints, res)
+	return
+}
+
+// partialLayer represents a subset of a layer (i.e., nodes at some
+// level in the Merkle tree).
+type partialLayer map[uint64]crypto.Digest
+
+// up takes a partial layer at level l, and returns the next-higher (partial)
+// level in the tree.  Since the layer is partial, up() requires siblings.
+//
+// The implementation is deterministic to ensure that up() asks for siblings
+// in the same order both when generating a proof, as well as when checking
+// the proof.
+//
+// If doHash is false, fill in zero hashes, which suffices for constructing
+// a proof.
+func (pl partialLayer) up(s *siblings, l uint64, doHash bool) (partialLayer, error) {
+	positions := make([]uint64, 0, len(pl))
+	for pos := range pl {
+		positions = append(positions, pos)
+	}
+	sort.Slice(positions, func(i, j int) bool { return positions[i] < positions[j] })
```

Comment:
If you want to avoid the sorting, you could change the `partialLayer` type to be a `[]IndexedHashable` where

```
type IndexedHashable struct {
    Index uint64
    Hash crypto.Hashable
}
```

Then you could kill the sorting, because as you append to the next `partialLayer`, things would stay sorted. Then you'd only have to sort the leaves layer up in `Verify` (or have callers to `Verify` pass a sorted slice of `IndexedHashable` -- you can verify that things are correctly sorted in the for loop below). To check for your sibling you'd just check the `Index` one after you in the partial layer.

I like the simplicity of maps but if we're optimizing for performance, this might be the way to go?

Reply:
Good point.  I switched to slices, which noticeably sped up both `Prove` and `Verify`.